Add STM32U5 support to STM32 MIPI DSI driver

CharlesDias opened the pull request:
Added STM32U5-specific configurations for the STM32 MIPI DSI driver, including:
- DSI PHY clock source setup.
- Logging for STM32U5-specific properties (e.g., PHY frequency range, low-power offset).
- Adjusted lane clock calculations for STM32U5 PLL configurations.
- Conditional handling for STM32U5-specific properties like PLL tuning, charge pump, and VCO range.

Tested with:
`west build -p -b stm32u5g9j_dk1 --shield st_lcd_dsi_mb1835 samples/drivers/display/`
And also
`west build -p -b stm32u5g9j_dk1 --shield st_lcd_dsi_mb1835 samples/modules/lvgl/demos -- -DCONFIG_LV_Z_DEMO_MUSIC=y `

Related to:
- https://github.com/zephyrproject-rtos/zephyr/pull/94420
- https://github.com/zephyrproject-rtos/zephyr/pull/94423

https://github.com/user-attachments/assets/a1ea4998-6c59-47e9-9aa0-e102e5e0c093

CharlesDias commented on dts/bindings/mipi-dsi/st,stm32-mipi-dsi.yaml:
Done

dsseng replied:
Well, as long as these properties represent hardware settings well enough (e.g. correspond to registers) these should work well even if e.g. the HAL is swapped to a pure register access.

My brief grep through the code of the HAL seems to indicate it loosely matches the registers.

"humanized" names of the registers seem well suited for DT: understandable without platform-specific knowledge, specify values of real things not abstractions

CharlesDias replied:
Thanks, all! Enlightening comments! 

About this _"So I think we should have a separate compatible for DSI Wrapper that will be SoC specific, and its node will be a child of DSI Host node in DT."_, any suggestions for the node name and binding file? Has ST implemented something similar?

CharlesDias commented on drivers/mipi_dsi/dsi_stm32.c:
```diff
@@ -56,6 +56,33 @@ struct mipi_dsi_stm32_data {
+		return -EIO;
+	}
+
+	uint32_t reg_val = *(volatile uint32_t *)0x46020CE4;
```
Done

CharlesDias commented on drivers/mipi_dsi/dsi_stm32.c:
```diff
@@ -56,6 +56,30 @@ struct mipi_dsi_stm32_data {
+	uint32_t dsi_sel = (__HAL_RCC_GET_DSI_SOURCE() == RCC_DSICLKSOURCE_DSIPHY);
+
+	LOG_DBG("DSI kernel clock source selection");
+	LOG_DBG("  RCC_CCIPR2_DSIHOSTSEL: %u", dsi_sel);
```
Done

CharlesDias commented:
Implemented the `DSIPHYInitPeriph` and `LOG` changes.

erwango commented on drivers/mipi_dsi/dsi_stm32.c:
```diff
@@ -56,6 +56,28 @@ struct mipi_dsi_stm32_data {
+	DSIPHYInitPeriph.PeriphClockSelection = RCC_PERIPHCLK_DSI;
+	DSIPHYInitPeriph.DsiClockSelection = RCC_DSICLKSOURCE_DSIPHY;
+
+	if (HAL_RCCEx_PeriphCLKConfig(&DSIPHYInitPeriph) != HAL_OK) {
+		LOG_ERR("Failed to configure DSI PHY as DSI clock source");
+		return -EIO;
+	}
```
Please use clock_control api when possible ?

CharlesDias replied:
Hi @erwango. Thank you for your review.

I believe that it is not possible to do this thought `clock_control`. I checked the `zephyr/drivers/` folder and I didn't find any calls to `HAL_RCCEx_PeriphCLKConfig`, `LL_RCC_SetDSIClockSource`, or `RCC_CCIPR2_DSIHOSTSEL`. Please let me know if I'm missing any point.

erwango replied:
This should be done using dts domain clock and add `dsisrc` as follows:
			clock-names = "dsiclk", "dsisrc", "refclk", "pixelclk";
			clocks = <&rcc STM32_CLOCK(APB2, 27U)>,
					<&rcc STM32_SRC_DSIPHY DSIHOST_SEL(1)>,
					<&rcc STM32_SRC_HSE NO_SEL>,
					<&rcc STM32_SRC_PLL3_R NO_SEL>;

Then, all `clock_control_configure()` in DSI driver.

Note that `STM32_SRC_DSIPHY` support might have to be added to U5 clock control dirver.
Also `DSIHOST_SEL` may need to be renamed to `DSI_SEL` to match CCIPR2 description. I'm not sure why this is not the case today.

erwango commented:
@CharlesDias No rush, but for my information, do you plan to complete this for next DV ?

JarmouniA commented on drivers/mipi_dsi/dsi_stm32.c:
```diff
@@ -212,13 +251,18 @@ static int mipi_dsi_stm32_host_init(const struct device *dev)
 
+#ifdef CONFIG_SOC_SERIES_STM32U5X
+#warning "HAL_DSI_SetLowPowerRXFilter is not implemented for STM32U5"
+#else
```
```suggestion
#ifndef CONFIG_SOC_SERIES_STM32U5X
```

CharlesDias commented on dts/bindings/mipi-dsi/st,stm32u5-mipi-dsi.yaml:
Hi @JarmouniA.

I'd like to confirm that the previous suggestion, discussed here https://github.com/zephyrproject-rtos/zephyr/pull/94433#discussion_r2277451605, was approved. Thanks!

CharlesDias commented on drivers/mipi_dsi/dsi_stm32.c:
```diff
@@ -212,13 +251,18 @@ static int mipi_dsi_stm32_host_init(const struct device *dev)
 
+#ifdef CONFIG_SOC_SERIES_STM32U5X
+#warning "HAL_DSI_SetLowPowerRXFilter is not implemented for STM32U5"
+#else
```
Thanks for your review! This will be removed as suggested on https://github.com/zephyrproject-rtos/zephyr/pull/94433#discussion_r2435866330.

erwango commented on drivers/mipi_dsi/dsi_stm32.c:
```diff
@@ -56,6 +56,28 @@ struct mipi_dsi_stm32_data {
 	uint32_t pixel_clk_khz;
 };
 
+#ifdef CONFIG_SOC_SERIES_STM32U5X
```
Prefer `#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32u5_mipi_dsi)`which provide the advantage to be compatible with the next soc which will use the same IP.

erwango replied:
> I haven't implemented it because I'd like to confirm if this change is to be done in all #ifdef CONFIG_SOC_SERIES_STM32U5X occurrences? 

Please do

CharlesDias replied:
Done

avolmat-st commented on drivers/mipi_dsi/dsi_stm32.c:
```diff
@@ -167,9 +200,15 @@ static int mipi_dsi_stm32_host_init(const struct device *dev)
+#ifdef CONFIG_SOC_SERIES_STM32U5X
+	/* LANE_BYTE_CLOCK = CLK_IN / PLLIDF * 2 * PLLNDIV / PLLODF / 8 */
+	data->lane_clk_khz = hse_clock / data->pll_init.PLLIDF * 2 * data->pll_init.PLLNDIV /
+			     data->pll_init.PLLODF / 8 / 1000;
```
This highlight a change in the way to define pll-odf between U5 and other platforms. Considering that it is defined the same way in the DT yaml, I don't think it should be used differently between the U5 series and other series.
(or it should be highlighted why this is needed, and highlight in the yaml).
Could you share you dts configuration of your mipi-dsi node for this U5 ?
I haven't check the U5 DSI spec, why lane clock freq calculation has to be different between series ?

avolmat-st replied:
Having checked the HAL code, my understanding is that PLLODF can have a value going from 1 to 9 (1 being divide by 1). So This seems to be ok as far as INIT is concerned, however the real divider factor a 1 << ODF

I am actually wondering if the original computation is correct ...

avolmat-st replied:
Thanks @CharlesDias for the clarification. I didn't notice that important different. It make senses then.
U5 support is ok I think then. What might have to be reworked is thus more the H7 rather than the U5 in fact since in case of H7, the PLLODF is not really giving the divider factor by the value to be written into the register, which is not the divider factor. But that is not part of this PR.

CharlesDias committed:
dts: bindings: mipi-dsi: extend STM32 MIPI DSI support

These additions enhance the flexibility of the MIPI DSI host configuration
for STM32U5 series, enabling finer control over the DSI PLL and PHY
settings.

Signed-off-by: Charles Dias <charlesdias.cd@outlook.com>

CharlesDias committed:
dts: arm: stm32u5: add MIPI DSI support for STM32U59x/5Ax series

Add DSI host controller node definition for STM32U59x/5Ax series.

Signed-off-by: Charles Dias <charlesdias.cd@outlook.com>

CharlesDias committed:
drivers: mipi_dsi: Add STM32U5 support to DSI driver

These changes enhance the driver's compatibility with the STM32U5 series,
enabling its use in applications requiring MIPI.

Signed-off-by: Charles Dias <charlesdias.cd@outlook.com>

CharlesDias commented on dts/bindings/mipi-dsi/st,stm32u5-mipi-dsi.yaml:
```diff
@@ -0,0 +1,40 @@
+
+description: |
+    STM32U5 MIPI DSI controller. This is a variant of the STM32 MIPI DSI
+    controller with specific properties for the U5 series.
```
done

CharlesDias commented on dts/bindings/mipi-dsi/st,stm32u5-mipi-dsi.yaml:
```diff
@@ -0,0 +1,40 @@
+    type: int
+    description: |
+      D-PHY PLL input frequency range. This is used to select the appropriate
+      frequency range for the D-PHY PLL operation on STM32U5 series.
```
done

CharlesDias commented:
Implemented the requests from @JarmouniA and @erwango (except for this one https://github.com/zephyrproject-rtos/zephyr/pull/94433#discussion_r2436092615).

Please verify if the dts/bindings/mipi-dsi/st,stm32u5-mipi-dsi.yaml is correct this way.

CharlesDias commented:
Implement conditional initialization of `dsisrc_clk`.

erwango commented:
> (except for this one https://github.com/zephyrproject-rtos/zephyr/pull/94433#discussion_r2436092615).

Can you explicit the reason ?

erwango commented on drivers/mipi_dsi/dsi_stm32.c:
```diff
@@ -457,6 +507,12 @@ static int mipi_dsi_stm32_init(const struct device *dev)
+				.enr = DT_INST_CLOCKS_CELL_BY_NAME(inst, dsisrc, bits),		\
+				.bus = DT_INST_CLOCKS_CELL_BY_NAME(inst, dsisrc, bus),		\
+			},),									\
+			(.dsisrc_clk = {0},))							\
```
Non blocking, for readability and better fit with other clocks
```suggestion
		.dsisrc_clk =  COND_CODE_1(DT_INST_CLOCKS_HAS_NAME(inst, dsisrc),		\
				({							\
				.enr = DT_INST_CLOCKS_CELL_BY_NAME(inst, dsisrc, bits),		\
				.bus = DT_INST_CLOCKS_CELL_BY_NAME(inst, dsisrc, bus),		\
			},),{0})
```

erwango commented on drivers/mipi_dsi/dsi_stm32.c:
```diff
@@ -185,46 +223,49 @@ static int mipi_dsi_stm32_host_init(const struct device *dev)
 	}
 
+#ifndef CONFIG_SOC_SERIES_STM32U5X
+
```
Non blocking:
```suggestion
#if !DT_HAS_COMPAT_STATUS_OKAY(st_stm32u5_mipi_dsi)
```
